pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -913,20 +905,21 @@ func (s *DiscoveryServer) startPush(version string, push *model.PushContext, ful
 				// This may happen to some clients if the other side is in a bad state and can't receive.
 				// The tests were catching this - one of the client was not reading.
 				pushTimeouts.Add(1)
-				client.mu.Lock()
-				if client.LastPushFailure.IsZero() {
-					client.LastPushFailure = time.Now()
-					client.mu.Unlock()
-					adsLog.Warnf("Failed to push, client busy %s", client.ConID)
-					pushErrors.With(prometheus.Labels{"type": "retry"}).Add(1)
-				} else if time.Since(client.LastPushFailure) > 10*time.Second {
-					client.mu.Unlock()
+				if lastPushFailure.IsZero() {
+					lastPushFailure = time.Now()
+
+				}
+				if time.Since(lastPushFailure) > 10*time.Second {
 					adsLog.Warnf("Repeated failure to push %s", client.ConID)
 					// unfortunately grpc go doesn't allow closing (unblocking) the stream.
 					pushErrors.With(prometheus.Labels{"type": "unrecoverable"}).Add(1)
 					pushTimeoutFailures.Add(1)
 					return
 				}
+
+				adsLog.Warnf("Failed to push, client busy %s", client.ConID)
+				pushErrors.With(prometheus.Labels{"type": "retry"}).Add(1)
+
```

Comment:
Why do you move this two lines out of the if branch?

Reply:
Sorry it wasn't clear. Before, the logic was wrong because if their was a timeout it only was counted the first time. I think w should count every timeout?

Reply:
The pushTimeout is 5s, and the retry timeout is 10s, so the retry count is only one time I think. So it doesn't matter.

Reply:
We start the timer after the first timeout though. So it times out, set timer to 0s. Times out again, now at 5s. Then when it times out the next time it's 10s and completely fails.